src/Ledger/TokenAlgebra/ValueRelation.lagda

```diff
@@ -0,0 +1,502 @@
+           ; _≈_ = _≋_
+           ; _∙_ = _⊕_
+           ; ε = ι , (λ 0b 0b' → trans 0b (sym 0b'))
+           ; isCommutativeMonoid = record { isMonoid = Vm ; comm = ⊕-comm }
+           }
+           where
+           isSemigrp : IsSemigroup _≋_ _⊕_
+           isSemigrp = record
+                         { isMagma = record
+                                       { isEquivalence = ≋-isEquivalence
+                                       ; ∙-cong = λ{u}{v}{w}{x} y z → ⊕-cong {u}{v}{w}{x} y z
+                                       }
+                         ; assoc = ⊕-assoc
+                         }
+
+           Vm : IsMonoid _≋_ _⊕_ (ι , (λ 0b 0b' → trans 0b (sym 0b')))
+           Vm = record { isSemigroup = isSemigrp ; identity = ι-identity }
+
+   mh : IsMonoidMorphism (monoid Vcm) (monoid +-0-commutativeMonoid) _hasHowMuchAda
+   IsSemigroupMorphism.⟦⟧-cong (IsMonoidMorphism.sm-homo mh) {R} {R'} R⇔R' = goal
+    where
+    q q' : Quantity
+    q = R hasHowMuchAda
+    q' = R' hasHowMuchAda
+
+    RR' : (proj₁ R) ⇐⇒ (proj₁ R')
+    RR' = Lift.lower R⇔R'
+
+    Rq : (proj₁ R) (AdaPolicy , AdaName) q
+    Rq = proj₂ (AdaForAll R)
+
+    R'q' : (proj₁ R') (AdaPolicy , AdaName) q'
+    R'q' = proj₂ (AdaForAll R')
+
+    Rq' : (proj₁ R) (AdaPolicy , AdaName) q'
+    Rq' = proj₂ RR' R'q'
+
+    goal : R hasHowMuchAda ≡ R' hasHowMuchAda
+    goal = proj₂ R Rq Rq'
+
+
+
+   IsSemigroupMorphism.∙-homo (IsMonoidMorphism.sm-homo mh) Ru Rv = goal
+    where
+    q qu qv : Quantity
+    q = (Ru ⊕ Rv) hasHowMuchAda
+    ξ : (proj₁ (Ru ⊕ Rv)) (AdaPolicy , AdaName) q
+    ξ = proj₂ (AdaForAll (Ru ⊕ Rv))
+    qu = Ru hasHowMuchAda
+    qv = Rv hasHowMuchAda
+
+    α : proj₁ Ru (AdaPolicy , AdaName) (proj₁ ξ)
+    α = proj₁ (proj₂ (proj₂ (proj₂ (AdaForAll (Ru ⊕ Rv)))))
+    β : proj₁ Rv (AdaPolicy , AdaName) (proj₁ (proj₂ ξ))
+    β = proj₁ (proj₂ (proj₂ (proj₂ (proj₂ (AdaForAll (Ru ⊕ Rv))))))
+
+
+    quξ : qu ≡ proj₁ ξ
+    quξ = proj₂ Ru (proj₂ (AdaForAll Ru)) α
+    qvξ : qv ≡ proj₁ (proj₂ ξ)
+    qvξ = proj₂ Rv (proj₂ (AdaForAll Rv)) β
+
+    γ : proj₁ ξ +ℕ proj₁ (proj₂ ξ) ≡ qu +ℕ qv
+    γ = begin
+         proj₁ ξ +ℕ proj₁ (proj₂ ξ)  ≡⟨ cong (λ x → proj₁ ξ + x) (sym qvξ) ⟩
+         proj₁ ξ +ℕ qv               ≡⟨ cong (λ x → x + qv ) (sym quξ) ⟩
+         qu +ℕ qv                    ∎
+
+    ξ' : q ≡ proj₁ ξ +ℕ proj₁ (proj₂ ξ)
+    ξ' = proj₂ (proj₂ (proj₂ (proj₂ ξ)))
+
+    goal : q ≡ qu +ℕ qv
+    goal = trans ξ' γ
+
+   IsMonoidMorphism.ε-homo mh = goal
+    where
+    I : AssetID ⇀ Quantity
+    I = ι , (λ 0b 0b' → trans 0b (sym 0b'))
+
+    q : Quantity
+    q = I hasHowMuchAda
+
+    dichlem : ∀{q : ℕ} → q ≡ 0ℕ ⊎ ¬(q ≡ 0ℕ)
+    dichlem {zero} = inj₁ _≡_.refl
+    dichlem {suc q} = inj₂ (λ ())
+
+    PQ : q ≡ 0ℕ ⊎ ¬(q ≡ 0ℕ)
+    PQ = dichlem
+
+    taut : (q ≡ 0ℕ) → (q ≡ 0ℕ)
+    taut = λ z → z
+
+    contr : ¬ (q ≡ 0ℕ) → (q ≡ 0ℕ)
+    contr np = ι-left-unique{AdaPolicy , AdaName} (proj₂ (AdaForAll I)) _≡_.refl
+
+    goal : q ≡ 0ℕ
```

Comment:
This can be simplified a lot by inlining & switching to `Dec`. Here's one option:
```
    goal : I hasHowMuchAda ≡ 0ℕ
    goal with I hasHowMuchAda ≟ 0ℕ
    ... | yes p = p
    ... | no _  = ι-left-unique{AdaPolicy , AdaName} (proj₂ (AdaForAll I)) _≡_.refl
```
Since the neutral element is also used multiple times, you could move that into a global `private` definition. Then you can even do
```
   IsMonoidMorphism.ε-homo mh with I hasHowMuchAda ≟ 0ℕ
   ... | yes p = p
   ... | no _  = ι-left-unique{AdaPolicy , AdaName} (proj₂ (AdaForAll I)) _≡_.refl
```

Reply:
Yes, thank you... good.  (I must have been a bit tired when I wrote that proof.)